crates/gitbutler-repo/tests/change_reference.rs

```diff
@@ -0,0 +1,254 @@
+    );
+    let last = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/last".into(),
+        test_ctx.commits.last().unwrap().change_id().unwrap(),
+    )?;
+    assert_eq!(last.branch_id, test_ctx.branch.id);
+    assert_eq!(last.name, "refs/remotes/origin/last".into());
+    assert_eq!(
+        last.change_id,
+        test_ctx.commits.last().unwrap().change_id().unwrap()
+    );
+    Ok(())
+}
+
+#[test]
+fn create_fails_with_non_remote_reference() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "foo".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    );
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        "Failed to parse the provided reference",
+    );
+    Ok(())
+}
+
+#[test]
+fn create_fails_when_branch_reference_with_name_exists() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/taken".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    )?;
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/taken".into(),
+        test_ctx.commits.last().unwrap().change_id().unwrap(),
+    );
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        format!("A reference refs/remotes/origin/taken already exists",),
+    );
+
+    Ok(())
+}
+
+#[test]
+fn create_fails_when_commit_already_referenced() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/one".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    )?;
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/two".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    );
+    assert_eq!(
+        result.unwrap_err().to_string(),
+        format!(
+            "A reference for change {} already exists",
+            test_ctx.commits.first().unwrap().change_id().unwrap()
+        ),
+    );
+
+    Ok(())
+}
+
+#[test]
+fn create_fails_when_commit_in_anothe_branch() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    let wrong_change = test_ctx.other_commits.first().unwrap().change_id().unwrap();
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/asdf".into(),
+        wrong_change.clone(),
+    );
+    assert!(result.is_err());
```

Comment:
One *could* even use `unwrap_err().to_string()` and assert on the error message.
Doing so of course increases the likelihood of test failures once the message changes, but also indicates that it is an important text for the user if it's expected to happen a lot.

For now, I'd also think that these errors aren't expected to be common though, so probably better to not assert on their text.